pkg/migration-tool/writer/writer.go

```diff
@@ -0,0 +1,168 @@
+package writer
+
+import (
+	"context"
+	"fmt"
+	"go.uber.org/atomic"
+	"time"
+
+	"github.com/gogo/protobuf/proto"
+	"github.com/golang/snappy"
+	"github.com/prometheus/common/model"
+	"github.com/prometheus/prometheus/prompb"
+	"github.com/prometheus/prometheus/storage/remote"
+	"github.com/timescale/promscale/pkg/log"
+	"github.com/timescale/promscale/pkg/migration-tool/planner"
+	"github.com/timescale/promscale/pkg/migration-tool/utils"
+)
+
+const (
+	defaultWriteTimeout  = time.Minute * 5
+	backOffRetryDuration = time.Millisecond * 100
+)
+
+type RemoteWrite struct {
+	c                  context.Context
+	sigBlockRead       chan *planner.Block
+	client             *utils.Client
+	blocksPushed       atomic.Int64
+	progressMetricName string // Metric name to main the last pushed maxt to remote write storage.
+	migrationJobName   string // Label value to the progress metric.
+	progressTimeSeries *prompb.TimeSeries
+}
+
+// New returns a new remote write. It is responsible for writing to the remote write storage.
+func New(c context.Context, remoteWriteUrl, progressMetricName, migrationJobName string, sigRead chan *planner.Block) (*RemoteWrite, error) {
+	wc, err := utils.CreateWriteClient(fmt.Sprintf("writer-%d", 1), remoteWriteUrl, model.Duration(defaultWriteTimeout))
+	if err != nil {
+		return nil, fmt.Errorf("creating write-client: %w", err)
+	}
+	write := &RemoteWrite{
+		c:                  c,
+		client:             wc,
+		sigBlockRead:       sigRead,
+		migrationJobName:   migrationJobName,
+		progressMetricName: progressMetricName,
+		progressTimeSeries: &prompb.TimeSeries{
+			Labels: utils.LabelSet(progressMetricName, migrationJobName),
+		},
+	}
+	write.blocksPushed.Store(0)
+	return write, nil
+}
+
+// Run runs the remote-writer. It waits for the remote-reader to give access to the in-memory
+// data-block that is written after the most recent fetch. After reading the block and storing
+// the data locally, it gives back the writing access to the remote-reader for further fetches.
+func (rw *RemoteWrite) Run(errChan chan<- error) {
+	var (
+		buf []byte
+		err error
+		ts  []prompb.TimeSeries
+	)
+	go func() {
+		defer func() {
+			log.Info("msg", "writer is down")
+			close(errChan)
+		}()
+		log.Info("msg", "writer is up")
+		for {
+			select {
+			case <-rw.c.Done():
+				break
+			case blockRef, ok := <-rw.sigBlockRead:
+				if !ok {
+					return
+				}
+				blockRef.SetDescription(fmt.Sprintf("pushing %.2f...", float64(blockRef.BytesCompressed())/float64(utils.Megabyte)), 1)
+				rw.progressTimeSeries.Samples = []prompb.Sample{{Timestamp: blockRef.Maxt(), Value: 1}} // One sample per block, else it will lead to duplicate samples.
+				blockRef.Timeseries = append(blockRef.Timeseries, rw.progressTimeSeries)
+				ts = timeseriesRefToTimeseries(blockRef.Timeseries)
```

Comment:
Can we encapsulate all this logic inside Block? Seems like there should be a function
```
ts := blockRef.GetTimeSeries(some arguments)
```
The arguments could be rw.ProgressTimeSeries. Then that function can set the description, set the sample in the progressTimeSeries, append the progress, convert to timeseries and return.